django/db/models/expressions.py

```diff
@@ -582,6 +593,126 @@ def get_group_by_cols(self):
+        self.condition = self._parse_expression(condition)
+        self.result = self._parse_expression(then)
+
+    def get_source_expressions(self):
+        return [self.condition, self.result]
+
+    def set_source_expressions(self, exprs):
+        self.condition, self.result = exprs
+
+    def get_source_fields(self):
+        """We're only interested in the fields of the result expressions."""
+        return [self.result._output_field_or_none]
+
+    def _parse_expression(self, expression):
+        return expression if hasattr(expression, 'resolve_expression') else F(expression)
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        c = self.copy()
+        c.is_summary = summarize
+        c.condition = c.condition.resolve_expression(query, allow_joins, reuse, summarize, False)
+        c.result = c.result.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        return c
+
+    def as_sql(self, compiler, connection, template=None):
+        template_params = {}
+        sql_params = []
+        condition_sql, condition_params = compiler.compile(self.condition)
+        template_params['condition'] = condition_sql
+        sql_params.extend(condition_params)
+        result_sql, result_params = compiler.compile(self.result)
+        template_params['result'] = result_sql
+        sql_params.extend(result_params)
+        template = template or self.template
+        return template % template_params, sql_params
+
+    def get_group_by_cols(self):
+        # This is not a complete expression, and cannot be used in GROUP BY
+        cols = []
+        for source in self.get_source_expressions():
+            cols.extend(source.get_group_by_cols())
+        return cols
+
+
+class Case(ExpressionNode):
+    """
+    An SQL searched CASE expression:
+
+        CASE
+            WHEN n > 0
+                THEN 'positive'
+            WHEN n < 0
+                THEN 'negative'
+            ELSE 'zero'
+        END
+    """
+    template = 'CASE %(cases)s ELSE %(default)s END'
+    case_joiner = ' '
+
+    def __init__(self, *cases, **extra):
+        default = extra.pop('default', Value(None))
+        output_field = extra.pop('output_field', None)
+        super(Case, self).__init__(output_field)
+        self._check_cases(cases)
+        self.cases = list(cases)
+        self.default = default if hasattr(default, 'resolve_expression') else F(default)
+
+    def _check_cases(self, cases):
+        for case in cases:
+            if not isinstance(case, When):
+                raise TypeError("Positional argument is not a When object.")
+            if not hasattr(case.condition, 'resolve_expression'):
+                raise TypeError("Positional argument is not an expression.")
+        if not all(isinstance(case.condition, Q) for case in cases):
+            raise TypeError("SearchedCase conditions must be lookups.")
```

Comment:
`SearchedCase` is no longer it's own type so this should be named `Case conditions must be lookups`. Alternatively, do this checking directly inside the `When` types, since users will be using these directly.